### PR TITLE
Cilium: route/ipsec, add mask to delete rule and add a set of unit tests

### PR DESCRIPTION
**Pull request by jrfastab**
This adds the mask to delete route in the IPv6 case to be sure we only delete routes that are expected. This was found while adding unit tests included in this series for routes and ipsec.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8176)

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 1]
```diff
@@ -335,6 +366,13 @@ func LoadIPSecKeysFile(path string) (int, uint8, error) {
 	return loadIPSecKeys(file)
 }
 
+// LoadIPSecKeys imports IPSec auth and crypt keys from a byte[]. The format
+// is a key per line as follows, ([id] auth-algo auth-key enc-algo enc-key)
+// Returns the authentication overhead in bytes, the key ID, and an error.
+func LoadIPSecKeys(keys []byte) (int, uint8, error) {
```
Comment (aanm) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:372]: @jrfastab it seems this function was only added for unit tests. Why not having it in the unit tests files?

**Review by aanm — CHANGES_REQUESTED** [review]
One tiny thing. LGTM otherwise.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 2]
```diff
@@ -335,6 +366,13 @@ func LoadIPSecKeysFile(path string) (int, uint8, error) {
...
+	k := bytes.NewReader(keys)
+	return loadIPSecKeys(k)
+}
```
Comment (joestringer) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:375]: nit: No newline after function

File: pkg/datapath/linux/node_linux_test.go [hunk 3]
```diff
@@ -1182,3 +1194,135 @@ func (s *linuxPrivilegedBaseTestSuite) BenchmarkNodeValidateImplementationDirect
...
+		c.Assert(encryptRule, check.Equals, true)
+
+		decryptRule, err := route.LookupDecryptIPv6Rule()
+		c.Assert(err, check.IsNil)
+		c.Assert(decryptRule, check.Equals, true)
+	}
+
+	// Delete node and verify encryption state and routes are removed
+	err = linuxNodeHandler.NodeDelete(nodev2)
+	c.Assert(err, check.IsNil)
```
Comment (joestringer) [inline pkg/datapath/linux/node_linux_test.go:1294]: Presumably after this, ciliumIP1 should still be present, to guarantee that the delete doesn't delete everything? Could be an extension to this test.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go [hunk 4]
```diff
@@ -69,6 +69,61 @@ func (p *IPSecSuitePrivileged) TestLoadKeys(c *C) {
...
+
+	err = ipSecReplacePolicyOut(local, remote)
+	c.Assert(err, IsNil)
+
+	xfrmStateList, err = netlink.XfrmStateList(0)
+	c.Assert(err, IsNil)
+	xfrmPolicyList, err = netlink.XfrmPolicyList(0)
+	c.Assert(err, IsNil)
+
+	c.Assert(stateLen+1, Equals, len(xfrmStateList))
```
Comment (joestringer) [inline pkg/datapath/linux/ipsec/ipsec_linux_test.go:112]: Why does replacing state result in a state list that's longer than the prior list?

**Review by joestringer — APPROVED** [review]
Some minor comments below. Assuming they're not major, LGTM.

**Comment by jrfastab** [comment]
marking WIP until I get some time to fix this and address comments

File: pkg/datapath/linux/ipsec/ipsec_linux.go (hunk 1 above)
Comment (jrfastab) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:372]: Right, sure we can add it to unit test files instead.

File: pkg/datapath/linux/ipsec/ipsec_linux_test.go (hunk 4 above)
Comment (jrfastab) [reply pkg/datapath/linux/ipsec/ipsec_linux_test.go:112]: Replace should add the state if it doesn't exist already and in this case it shouldn't exist in the {State|Policy}List

Comment (jrfastab) [reply pkg/datapath/linux/ipsec/ipsec_linux_test.go:112]: Replace should add the state if it doesn't exist already and in this case it shouldn't exist in the {State|Policy}List

**Pull request closed without merging**